prometheus/metrics.go

```diff
@@ -72,37 +72,59 @@ type Histogram struct {
 	Percentile99 *Metric `json:"percentile99"`
 }
 
-// Returns <healthy, total, error>
-func getServiceHealth(api v1.API, namespace string, servicename string) (int, int, error) {
+// EnvoyHealth is the number of healthy versus total membership (ie. replicas) inside envoy cluster (ie. service)
+type EnvoyHealth struct {
+	Inbound  EnvoyRatio `json:"inbound"`
+	Outbound EnvoyRatio `json:"outbound"`
+}
+
+// EnvoyRatio is the number of healthy members versus total members
+type EnvoyRatio struct {
+	Healthy int `json:"healthy"`
+	Total   int `json:"total"`
+}
+
+func getServiceHealth(api v1.API, namespace string, servicename string) (EnvoyHealth, error) {
 	envoyClustername := strings.Replace(config.Get().ExternalServices.Istio.IstioIdentityDomain, ".", "_", -1)
 	queryPart := replaceInvalidCharacters(fmt.Sprintf("%s_%s_%s", servicename, namespace, envoyClustername))
 	now := time.Now()
+	ret := EnvoyHealth{}
 
-	var healthyVect, totalVect model.Vector
-	var healthyErr, totalErr error
-	var wg sync.WaitGroup
-	wg.Add(2)
-	go func() {
-		defer wg.Done()
-		healthyVect, healthyErr = fetchTimestamp(api, fmt.Sprintf("envoy_cluster_out_%s_http_membership_healthy", queryPart), now)
-	}()
+	// Note: metric names below probably depend on some istio configuration.
+	// They should anyway change soon in a more prometheus-friendly way,
+	// see https://github.com/istio/istio/issues/4854 and https://github.com/istio/istio/pull/5069
 
-	go func() {
-		defer wg.Done()
-		totalVect, totalErr = fetchTimestamp(api, fmt.Sprintf("envoy_cluster_out_%s_http_membership_total", queryPart), now)
-	}()
-
-	wg.Wait()
-
-	if healthyErr != nil {
-		return 0, 0, healthyErr
-	} else if totalErr != nil {
-		return 0, 0, totalErr
-	} else if len(healthyVect) == 0 || len(totalVect) == 0 {
-		// Missing metrics
-		return 0, 0, nil
+	// Inbound
+	vec, err := fetchTimestamp(api, fmt.Sprintf("envoy_cluster_inbound_9080__%s_membership_healthy", queryPart), now)
```

Comment:
All these queries are sequencial ?

Reply:
Yes I mentioned that in the PR description: remove some goroutines from fine-grained metric-fetching and add some to coarser level => this should optimize waiting time

There's 2 motivations for that. First is that if the threads are spawned earlier, each thread will be able to continue to run while others are waiting for the prometheus call. So it should be globally less waiting time.
And secondly I didn't want to make goroutines at both places in order to not generate too many threads. They still have a cost even in go :)

Reply:
Ok, I won't argue, as I'm not a fan of early performance optimizations.
But it was curious this change, because, even if you spawn in a higher level, I guess it should be spawn at this level too.
We were discussing something similar in the checkers logic and we introduced there some similar concurrency design, just to maximize the concurrent calls.
But anyway, that is something it can be iterated, perhaps the tradeoff is just minimal.
I don't have more data to discuss, just theory.

Reply:
Well wait a moment. I'll put them in both places. Kiali is not yet _that huge_ to worry so early about having too many threads.